core/src/main/scala/codesearch/core/index/repository/SourceRepository.scala

```diff
@@ -0,0 +1,48 @@
+package codesearch.core.index.repository
+import java.io.File
+import java.net.URI
+import java.nio.file.Path
+
+import org.apache.commons.io.FilenameUtils.getExtension
+
+import sys.process._
+import scala.concurrent.Future
+import scala.concurrent.ExecutionContext.Implicits.global
+
+object SourceRepository {
+  implicit def packageDownloader[A <: SourcePackage]: Download[A] = new Download[A] {
+    override def downloadSources(pack: A): Future[Int] = {
+      for {
+        archive   <- download(pack.url, pack.fsPath)
+        directory <- unzip(archive)
+        countArch <- deleteExcessFiles(archive, pack.extentions)
+        countDir  <- deleteExcessFiles(directory, pack.extentions)
+      } yield directory
+    }.map(_ => 1).recover { case _ => 0 }
+
+  }
+
+  def download(from: URI, to: Path): Future[File] = {
+    Future {
+      Seq("curl", "-o", to.toString, from.getPath) !!;
+      to
+    }
+    ???
+  }
+
+  def unzip(file: File): Future[File] = {
+    ???
+  }
+
+  def deleteExcessFiles(directory: File, allowedExtentions: Set[String]): Future[Int] = Future {
```

Comment:
Comments please

Reply:
The function filters files leaving only those whose extensions contains in allowable set.

Reply:
Comments in code, not on Github :slightly_smiling_face: